use buildcatrust to generate trust stores

lukegb opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

* Generate three types of trust store
* Support more methods of passing extra CAs to trust in a NixOS system
* Reenable p11-kit support in Firefox

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

lukegb committed:
buildcatrust: init at 0.1.3

lukegb committed:
cacert: port to use buildcatrust

This introduces the ability to have additional certificates in the trust
store using an override, similar to how the blacklist is done. If the
certificates are provided in OpenSSL TRUSTED CERTIFICATE form, then
those trust bits will be respected.

It also adds a p11-kit compatible trust store output.

lukegb committed:
p11-kit: look in /etc/ssl/trust-source directory as well

lukegb committed:
nixos/ca: use cacert package build for options and p11-kit output

The cacert package can now generate p11-kit-compatible output itself,
as well as generating the correct set of outputs for fully-joined
and unbundled "traditional" outputs (in standard PEM and
OpenSSL-compatible formats).

lukegb committed:
nixos/tests/custom-ca: reenable Firefox test

andir commented on nixos/modules/security/ca.nix:
```diff
@@ -8,12 +8,9 @@ let
 
   cacertPackage = pkgs.cacert.override {
     blacklist = cfg.caCertificateBlacklist;
+    extraCertificates = cfg.certificateFiles ++ cfg.certificates;
```
So this means we will require the full NSS source on every single machine that sets  a custom certificate? How much is it these days?

andir replied:
The answer is about 80MB gzip compressed. Could we introduce an intermediary step that extracts the files that we need for this and only use that as src here? That would allow downloading a miminal set of files (compressd) from the binary cache vs having to obtain the full source archive.

lukegb replied:
We could probably change this to take the output of the cacert package and then filter/add to _that_ instead; then if you block a CA it has the same cost as adding a new one.

lukegb replied:
I did the simpler thing you suggested because that's less effort ;)

lukegb commented on pkgs/development/python-modules/buildcatrust/default.nix:
```diff
@@ -0,0 +1,32 @@
+    pytestCheckHook
+  ];
+  disabledTestPaths = [
+    "buildcatrust/tests/test_nonhermetic.py"
```
Done.

lukegb commented on pkgs/development/python-modules/buildcatrust/default.nix:
```diff
@@ -0,0 +1,32 @@
+{ lib
+, buildPythonPackage
+, fetchPypi
+, pythonOlder
```
Oops. Removed.

lukegb commented:
Aaand I footgunned.

sternenseemann commented on pkgs/data/misc/cacert/default.nix:
```diff
@@ -1,123 +1,211 @@
 }:
 
+# extraCertificates should be paths or strings.
+assert lib.length (lib.filter (f: !(builtins.isPath f || builtins.isString f)) extraCertificates) == 0;
```
We probably want to allow results of `runCommand` or `writeText` as well?

```suggestion
assert lib.length (lib.filter (f: !(builtins.isPath f || builtins.isString f || lib.isDerivation f)) extraCertificates) == 0;
```

lukegb replied:
I'm going to rip this out, since I have those anyway in the NixOS module and this is needlessly complicated.

sternenseemann commented on pkgs/data/misc/cacert/default.nix:
```diff
@@ -1,123 +1,211 @@
-      "https://git.launchpad.net/ubuntu/+source/ca-certificates/plain/mozilla/certdata2pem.py?id=47e49e1e0a8a1ca74deda27f88fe181191562957"
-    ];
-    sha256 = "1d4q27j1gss0186a5m8bs5dk786w07ccyq0qi6xmd2zr1a8q16wy";
+  isPathIntoStore = f: builtins.isString f && lib.substring 0 (lib.stringLength builtins.storeDir) f == builtins.storeDir;
```
You should just be able to use `lib.isStorePath`

lukegb replied:
Alas. Store Paths are decreed to be the literal paths `/nix/store/fooo-hash` only, and not, for instance, any files that would happen to be under directories named thusly.

This also matches /nix/store itself, though, which is probably undesirable.

sternenseemann commented on pkgs/data/misc/cacert/default.nix:
```diff
@@ -1,123 +1,211 @@
-    sha256 = "1d4q27j1gss0186a5m8bs5dk786w07ccyq0qi6xmd2zr1a8q16wy";
+  isPathIntoStore = f: builtins.isString f && lib.substring 0 (lib.stringLength builtins.storeDir) f == builtins.storeDir;
+  extraCertificateFiles = lib.filter (f: builtins.isPath f || isPathIntoStore f) extraCertificates;
+  extraCertificateStrings = lib.filter builtins.isString extraCertificates;
```
I think this needs something like this:

```suggestion
  extraCertificateStrings = lib.filter (f: builtins.isString f && !(lib.isStorePath f)) extraCertificates;
```

Otherwise we'll mistake values like `"${foo}/bar.crt` for certificate strings.

sternenseemann commented on pkgs/data/misc/cacert/default.nix:
```diff
@@ -1,123 +1,211 @@
+      --p11kit_output ca-bundle.trust.p11-kit
   '';
 
   installPhase = ''
```
A bit surprised that nothing here uses `install(1)`, but eh.

lukegb replied:
Now it does. Much nicer! :P

sternenseemann reviewed (commented):
Initial impression, will have a more thorough look later hopefully.

lukegb commented:
nixos/tests/custom-ca.nix passes! (or well, it did before my most recent rebase on top of staging)

sternenseemann commented:
@ofborg build cacert
@ofborg test custom-ca

rnhmjoj commented on pkgs/data/misc/cacert/default.nix:
```diff
@@ -31,93 +22,160 @@ stdenv.mkDerivation rec {
-    mkdir -pv $unbundled/etc/ssl/certs
-    cp -v *.crt $unbundled/etc/ssl/certs
-    rm $unbundled/etc/ssl/certs/ca-bundle.crt  # not wanted in unbundled
+    install -D -t $unbundled/etc/ssl/certs unbundled/*.crt
```
```suggestion
    install -D -t "$unbundled/etc/ssl/certs" unbundled/*.crt
```
Even more quoting.

rnhmjoj commented on pkgs/data/misc/cacert/default.nix:
```diff
@@ -31,93 +22,160 @@ stdenv.mkDerivation rec {
+    install -D ca-bundle.crt $out/etc/ssl/certs/ca-bundle.crt
+
+    # install p11-kit specific output to p11kit output
+    install -D ca-bundle.trust.p11-kit $p11kit/etc/ssl/trust-source/ca-bundle.trust.p11-kit
```
```suggestion
    install -D ca-bundle.trust.p11-kit "$p11kit/etc/ssl/trust-source/"
```
Quoting again.

rnhmjoj commented on pkgs/data/misc/cacert/default.nix:
```diff
@@ -31,93 +22,160 @@ stdenv.mkDerivation rec {
   installPhase = ''
-    mkdir -pv $out/etc/ssl/certs
-    cp -v ca-bundle.crt $out/etc/ssl/certs
+    install -D ca-bundle.crt $out/etc/ssl/certs/ca-bundle.crt
```
```suggestion
    install -D ca-bundle.crt "$out/etc/ssl/certs/"
```
Variable should always be quoted, better safe than sorry.

rnhmjoj commented on pkgs/data/misc/cacert/default.nix:
```diff
@@ -31,93 +22,160 @@ stdenv.mkDerivation rec {
-    cat "${certdata2pem}" > certdata2pem.py
+    cat << EOF > extra-certificates-bundle.crt
+    ${lib.concatStringsSep "\n\n" extraCertificateStrings}
+    EOF
```
Same as above.

rnhmjoj commented on nixos/tests/custom-ca.nix:
```diff
@@ -153,17 +151,21 @@ in
+      "firefox",
       "chromium",
-      "falkon",
+      # "falkon",  # Falkon is broken because kwallet is broken.
```
This is ironic: I had picked falkon because I was assuming it was the most stable.
You should be able to replace it with `qutebrowser` and the error text is `Certificate error`.

rnhmjoj commented on nixos/tests/custom-ca.nix:
```diff
@@ -153,17 +151,21 @@ in
       "midori"
     ]
-    errors = ["Security Risk", "not private", "Certificate Error", "Security"]
+    errors = {
```
Good idea!

rnhmjoj reviewed (commented):
Thank you for working on this!
I left some comments and suggestions, mostly about shell scripting.

rnhmjoj commented:
I'd like to have this in the 21.11 release, before the feature freeze.
@lukegb do you think you have time to finish it? Otherwise I'll try to pick up the work and open another PR.

sternenseemann reviewed (changes requested):
Nothing else stood out to me on this pass.

lukegb committed:
cacert: extract certdata.txt from main package

This allows users to specify custom CAs without needing to download the
entirety of the NSS source code - just certdata.txt, which should end up
in cache.nixos.org.

mweinelt reviewed (approved):
Tested locally and Firefox looks good.

evrim commented:
Um, what about `security.certerrors.mitm.auto_enable_enterprise_roots` ?

https://support.mozilla.org/en-US/kb/how-disable-enterprise-roots-preference
